x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTrunc.java

```diff
@@ -0,0 +1,280 @@
+import org.elasticsearch.xpack.sql.expression.function.scalar.BinaryScalarFunction;
+import org.elasticsearch.xpack.sql.expression.gen.pipeline.Pipe;
+import org.elasticsearch.xpack.sql.expression.gen.script.ScriptTemplate;
+import org.elasticsearch.xpack.sql.tree.NodeInfo;
+import org.elasticsearch.xpack.sql.tree.Source;
+import org.elasticsearch.xpack.sql.type.DataType;
+import org.elasticsearch.xpack.sql.util.StringUtils;
+
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.time.temporal.ChronoField;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import static org.elasticsearch.common.logging.LoggerMessageFormat.format;
+import static org.elasticsearch.xpack.sql.expression.TypeResolutions.isDate;
+import static org.elasticsearch.xpack.sql.expression.TypeResolutions.isString;
+import static org.elasticsearch.xpack.sql.expression.function.scalar.datetime.DateTruncProcessor.process;
+import static org.elasticsearch.xpack.sql.expression.gen.script.ParamsBuilder.paramsBuilder;
+
+public class DateTrunc extends BinaryScalarFunction {
+
+    public enum DatePart {
+
+        MILLENNIUM("millennia"),
+        CENTURY("centuries"),
+        DECADE("decades"),
+        YEAR("years", "yy", "yyyy"),
+        QUARTER("quarters", "qq", "q"),
+        MONTH("months", "mm", "m"),
+        WEEK("weeks", "wk", "ww"),
+        DAY("days", "dd", "d"),
+        HOUR("hours", "hh"),
+        MINUTE("minutes", "mi", "n"),
+        SECOND("seconds", "ss", "s"),
+        MILLISECOND("milliseconds", "ms"),
+        MICROSECOND("microseconds", "mcs"),
+        NANOSECOND("nanoseconds", "ns");
+
+        private static final Set<String> ALL_DATE_PARTS;
+        private static final Map<String, DatePart> RESOLVE_MAP;
+
+        static {
+            ALL_DATE_PARTS = new HashSet<>();
+            RESOLVE_MAP = new HashMap<>();
+
+            for (DatePart datePart : DatePart.values()) {
+                String lowerCaseName = datePart.name().toLowerCase(IsoLocale.ROOT);
+                ALL_DATE_PARTS.add(lowerCaseName);
+                ALL_DATE_PARTS.addAll(datePart.aliases);
+
+                RESOLVE_MAP.put(lowerCaseName, datePart);
+                for (String alias : datePart.aliases) {
+                    RESOLVE_MAP.put(alias, datePart);
+                }
+            }
+        }
+
+        private Set<String> aliases;
+
+        DatePart(String... aliases) {
+            this.aliases = Set.of(aliases);
+        }
+
+        public static DatePart resolveTruncate(String truncateTo) {
+            return RESOLVE_MAP.get(truncateTo.toLowerCase(IsoLocale.ROOT));
+        }
+
+        public static List<String> findSimilar(String match) {
+            return StringUtils.findSimilar(match, ALL_DATE_PARTS);
+        }
+
+        public static ZonedDateTime truncate(ZonedDateTime dateTime, DateTrunc.DatePart datePart) {
```

Comment:
In my previous comment I was suggesting to move this extraction inside the enum itself so each one would have its own `of` implementation : `DateTruncate.NANOSECOND.of(ZonedDateTime)`